grails-app/services/com/unifina/service/StreamService.groovy

```diff
@@ -189,14 +182,9 @@ class StreamService {
      * @return Autocreated Stream field config as a Map (can be written to stream.config as JSON)
      */
 	@CompileStatic
-	public Map importCsv(CSVImporter csv, Stream stream) {
-		/**
-		 * Batch-imported rows have negative Kafka offsets to avoid collisions with messages actually produced
-		 * to the stream via Kafka.
-		 */
-		List<Long> latestOffsetByPartition = (0..stream.getPartitions()-1).collect { Integer partition ->
-			StreamrBinaryMessageWithKafkaMetadata msg = cassandraService.getLatestBeforeOffset(stream, partition, 0)
-			return msg ? msg.offset : MINIMUM_SAFE_INTEGER_IN_JAVASCRIPT
+	public Map importCsv(CSVImporter csv, Stream stream, String publisherId) {
+		List<StreamMessage> latestMessageByPartition = (0..stream.getPartitions()-1).collect { Integer partition ->
```

Comment:
In this method, thanks to the new message ids, I think there's basically no longer a need to know what messages pre-exist or maintain the `latestMessageByPartition` lookup. You can simply have a `long sequenceNumber = 0` and then `sequenceNumber++` after every `saveMessage`.

Reply:
This would work yes. But wasn't the initial logic supposed to order by timestamp and only use the sequence number for conflicting timestamps?

Reply:
> But wasn't the initial logic supposed to order by timestamp and only use the sequence number for conflicting timestamps?

Sure, the minimum requirement is that sequence numbers are unique within a timestamp. However fulfilling the stronger property of being unique within a session / csv file / universally does fulfill the minimum requirement plus additionally provide more information which may be useful for human debugging, estimating number of messages in a gap, or similar. Should we establish a best practice that exceeds the minimum requirement, or is it just confusing?

> Also how would we compute the previous timestamp and sequence number for each message?

It's not strictly necessary, it may be better to intentionally break the ref chain between bulk uploads, because they might happen in whatever order. Consider for example this scenario:

- There's existing data up to timestamp t1
- User uploads a batch of data from t3 to t4. First message in batch (t3) would now point to latest existing message (with timestamp t1)
- User uploads another batch of data from t2-t3. First message in batch would point to the same pre-existing message t1. Now both t3 and t2 would have a previousRef of t1.

Historical data uploads present other pretty annoying problems too, for example if the user uploads overlapping CSV files. Let's say file A has data from timestamp t1 to t3 and file B has data from t2 to t4. This leaves us with two parallel message chains, which would be an invalid situation. We could forbid this situation by checking that the timestamp range of the file contains no data in the stream.